boa/src/syntax/lexer/number.rs

```diff
@@ -0,0 +1,351 @@
+                        // Finish lexing number.
+
+                        // return Err(Error::syntax(
+                        //     "Found BigIntLiteralSuffix after non-integer number",
+                        // ));
+                    }
+                    Some(Ok('.')) => {
+                        // Found second . within decimal number
+                        // Finish lexing number.
+
+                        // return Err(Error::syntax("Found second . within decimal number"));
+                    }
+                    Some(Ok('e')) | Some(Ok('E')) => {
+                        // Consume the ExponentIndicator.
+                        cursor.next();
+
+                        take_signed_integer(&mut exp_str, cursor, &kind)?;
+                    }
+                    Some(Err(_e)) => {
+                        // todo!();
+                    }
+                    Some(Ok(_)) | None => {
+                        // Finished lexing.
+                        kind = NumericKind::Rational;
+                    }
+                    _ => {}
+                }
+            }
+            Some(Ok('e')) | Some(Ok('E')) => {
+                // Consume the ExponentIndicator.
+                cursor.next();
+
+                // buf.push('e');
+
+                take_signed_integer(&mut exp_str, cursor, &kind)?;
+            }
+            Some(Err(_e)) => {
+                // todo!();
+            }
+
+            Some(Ok(_)) | None => {
+                // Indicates lexing finished.
+            }
+        }
+
+        // unimplemented!();
+
+        // self.check_after_numeric_literal()?;
+
+        let num = match kind {
+            NumericKind::BigInt(base) => {
+                Numeric::BigInt(
+                    BigInt::from_string_radix(&buf, base as u32).expect("Could not convert to BigInt")
+                    )
+            }
+            NumericKind::Rational /* base: 10 */ => {
+                let r = f64::from_str(&buf).map_err(|_| Error::syntax("Could not convert value to f64"))?;
+                if exp_str == "" {
+                    Numeric::Rational(
+                        r
+                    )
+                } else {
+                    let n = f64::from_str(&exp_str).map_err(|_| Error::syntax("Could not convert value to f64"))?;
+                
+                    Numeric::Rational(
+                        r * f64::powf(10.0, n)
+                    )
+                }
+            }
+            NumericKind::Integer(base) => {
+                if let Ok(num) = i32::from_str_radix(&buf, base as u32) {
+                    if exp_str == "" {
+                        Numeric::Integer(num)
+                    } else {
+                        let n = i32::from_str(&exp_str).map_err(|_| Error::syntax("Could not convert value to f64"))?;
+                    
+                        Numeric::Integer(
+                            num * i32::pow(10, n as u32)
+                        )
```

Comment:
Here I return an integer (if possible) even if exponent notation is used. 

Previously a rational was returned but I believe an integer is more performant - thoughts?

Reply:
Actually disregard this is probably a bad idea as it might not fit.

Reply:
Actually, I think it's a very good idea, but we need to check the arithmetic by using `checked_mul()` and `checked_pow()`.

Then, I would use `try_from()` if possible to convert the exponent. If everything is ok, then, generate an integer, if not, use floating point arithmetics.

Reply:
Done in recent commit